### PR TITLE
Correctly check for presence of list view sort header.

### PR DESCRIPTION
**Pull request by ThFabba**
See https://github.com/reactos/reactos/commit/be9a0afc1f1b24a3edbd14dffe3b04814e85074c#commitcomment-25230680

### CODE REVIEW

File: base/applications/msconfig_new/comctl32ex/listview.c [hunk 1]
```diff
@@ -49,7 +49,7 @@ ListView_SortEx(HWND hListView,
     BOOL bSortAsc;
     Sort sort;
 
-    if ((GetWindowLongPtr(hListView, GWL_STYLE) & ~LVS_NOSORTHEADER) == 0)
+    if (GetWindowLongPtr(hListView, GWL_STYLE) & LVS_NOSORTHEADER)
```
Comment (HBelusca) [inline base/applications/msconfig_new/comctl32ex/listview.c]: ok

File: base/applications/regedit/listview.c [hunk 2]
```diff
@@ -478,7 +478,7 @@ static int CALLBACK CompareFunc(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSor
 
 static BOOL ListView_Sort(HWND hListView, int iSortingColumn, int iSortedColumn)
 {
-    if ( (GetWindowLongPtr(hListView, GWL_STYLE) & ~LVS_NOSORTHEADER) &&
+    if (!(GetWindowLongPtr(hListView, GWL_STYLE) & LVS_NOSORTHEADER) &&
```
Comment (HBelusca) [inline base/applications/regedit/listview.c]: ok

File: dll/win32/shell32/CDefView.cpp [hunk 3]
```diff
@@ -698,7 +698,7 @@ BOOL CDefView::_Sort()
     HWND hHeader;
     HDITEM hColumn;
 
-    if ((m_ListView.GetWindowLongPtr(GWL_STYLE) & ~LVS_NOSORTHEADER) == 0)
+    if (m_ListView.GetWindowLongPtr(GWL_STYLE) & LVS_NOSORTHEADER)
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp]: good to go!

**Comment by yagoulas** [comment]
Why do these checks exist in the first place?

**Comment by HBelusca** [comment]
To not perform any sorting in case the listview style changes (for example, we go from report mode with headers to e.g. icon view; or because an external module changes the list style).

**Comment by yagoulas** [comment]
You mean that we don't want the items of the shell to be sortable in the icon view?

**Comment by yagoulas** [comment]
And a second note. The interface of CDefView is so absurdly big because it handles all sorts of things that can be done on the underlying listview. Other modules messing with the encapsulated list view just isn't expected. If you want to hack around enapsulation you can, but it doesn't mean that you should consider this something that can happen.

**Comment by HBelusca** [comment]
> You mean that we don't want the items of the shell to be sortable in the icon view?
Actually I don't know. It would be nice they could be sorted wrt. "some criterium" (which I wonder how to choose since in non-report mode there is no column header, at least visible).
The point is, that the rest of the code heavily relies on the existence of the HEADER sub-control inside the listview:
```
hHeader = (HWND)m_ListView.SendMessage(LVM_GETHEADER, 0, 0);
```
I'm not sure that kind of thing would work OK when the listview is not in "report" mode.
Or we can check whether that hHeader could be NULL and quit the sort function if so.

**Comment by HBelusca** [comment]
Concerning your second comment:
> [...] The interface of CDefView is so absurdly big [...]

while this is true, I didn't add this sorting procedure e.g. inside the roscontrols CListView file because, it may be that one doesn't always want this sorting behaviour by default, or want a different behaviour there.

**Comment by yagoulas** [comment]
The listview has a header as long as there is a column inserted by the CDefView. When it is in the icon mode the header still exists but it is hidden. This allows sorting items to work even in the icon mode. And if no column was inserted then there would be nothign int the context menu to call this. Perhaps it needs to check if hHeader is NULL and then bail out. However you still don't know what you are doing. If you look at the implementation of our listview you will see that the only thing LVS_NOSORTHEADER does it change the style of the header control that would be created anyway.

**Pull request merged**